framework/Optimizers/OptimizerBase.py

```diff
@@ -0,0 +1,515 @@
+      self.optVars[traj]            = self.getOptVars(full=True) #initial as full space
+      self.submissionQueue[traj]    = deque()
+
+    #check initial point array consistency
+    rightLen = len(self.optTraj) #the hypothetical correct length
+    for var in self.getOptVars(full=True):
+      haveLen = len(self.optVarsInit['initial'][var])
+      if haveLen != rightLen:
+        self.raiseAnError(RuntimeError,'The number of trajectories for variable "{}" is incorrect!  Got {} but expected {}!  Check the <initial> block.'.format(var,haveLen,rightLen))
+
+    # extend multivalue variables (aka vector variables, or variables with "shape")
+    ## TODO someday take array of initial values from a DataSet
+    for var,shape in self.variableShapes.items():
+      if np.prod(shape) > 1:
+        for traj in self.optTraj:
+          baseVal = self.optVarsInit['initial'][var][traj]
+          if len(baseVal) == 1:
+            newVal = np.ones(shape)*baseVal
+            self.optVarsInit['initial'][var][traj] = newVal
+          elif len(baseVal) != np.prod(shape):
+            self.raiseAnError(IOError, "The number of initial values provided for vector variable", var," is not equal the vector total size!" )
+
+    # check the constraint here to check if the initial values violate it
+    varK = {}
+    for trajInd in self.optTraj:
+      for varName in self.getOptVars():
+        varK[varName] = self.optVarsInit['initial'][varName][trajInd]
+        self.checkConstraint(varK)
+
+    self.localInitialize(solutionExport=solutionExport)
+
+  def amIreadyToProvideAnInput(self):
+    """
+      This is a method that should be called from any user of the optimizer before requiring the generation of a new
+      input. This method act as a "traffic light" for generating a new input.
+      Reason for not being ready could be for example: exceeding number of model evaluation, convergence criteria met, etc.
+      @ In, None
+      @ Out, ready, bool, indicating the readiness of the optimizer to generate a new input.
+    """
+    ready = True if self.counter['mdlEval'] < self.limit['mdlEval'] else False
+    if not ready:
+      self.raiseAMessage('Reached limit for number of model evaluations!')
+    ready = self.localStillReady(ready)
+    return ready
+
+  def _incrementCounter(self):
+    """
+      Increments counter and sets up prefix.
+      @ In, None
+      @ Out, None
+    """
+    #since we are creating the input for the next run we increase the counter and global counter
+    self.counter['mdlEval'] += 1
+    self.inputInfo['prefix'] = str(self.counter['mdlEval'])
+
+  def updateVariableHistory(self,data,traj=0):
+    """
+      Stores new historical points into the optimization history.
+      @ In, data, dict, new input space entries as {var:#, var:#}
+      @ In, traj, int, integer label of the current trajectory of interest
+      @ Out, None
+    """
+    self.optVarsHist[traj][self.counter['varsUpdate'][traj]] = copy.deepcopy(data)
```

Comment:
I don't see optVarsHist in the `__init__` list. Also it looks like it might be nearly the same as the similar entry in `self.counter`; is there duplicate behavior?